### ISSUE TITLE
High memory allocation in prometheus.MakeLabelPair

### ISSUE THREAD

**Issue by jkroepke** [prometheus/client_golang#1702]
While investigate memory leak with windows_exporter, I figure out that there is a decent memory allocation rate in prometheus.MakeLabelPairs.

[windows_exporter](https://github.com/prometheus-community/windows_exporter) (like node_exporter) generate a high amount of ConstMetrics on each scrape call. The process collector expose around 6200 metrics on each call, resulting in 6200 MustNewConstMetric calls on each scrape.

I ask myself, if this can be optimized somehow. The heap dump is generate after 1 days of runtime, with 5 seconds scrape interval (to simulate a long running scenario)

the `dto.LabelPair` struct contains other struct related to protobuf. Not sure, if they needed. [Prometheus 2.0 does not support protobuf metris scrape](https://github.com/prometheus/client_model/blob/master/README.md), it feels like an expensive leftover. 

---

Flame Graph:
![2024-12-25 18_49_20-20 101 113 87_3389 - Remotedesktopverbindung](https://github.com/user-attachments/assets/c799fa97-cfd9-4d58-bc01-af81a7bc81b1)

Source View:
![2024-12-25 18_51_46-20 101 113 87_3389 - Remotedesktopverbindung](https://github.com/user-attachments/assets/23113855-3017-47ea-825b-8d91924dc8b0)

Heap Dump via pprof: [memory_profile_windows.pb.gz](https://github.com/user-attachments/files/18246848/memory_profile_windows.pb.gz)

**Comment by bwplotka** [comment]
Yup, that's a known problem for the `Const*` metric users. A great showcase is in [cadvisor](https://github.com/google/cadvisor/blob/be773c0a2b04536a818f79339bf89c82a74abb02/metrics/prometheus.go#L1879) also impacted by this. Kube state metrics get rid of client_golang for this reason for the scrape endpoints.

At some point I tried to optimize cadvisor code to essentially cache a metrics and update them in place or so, which is not trivial for the bigger codebase: https://github.com/google/cadvisor/pull/2974

There's definitely room to either make mutable caching layer for bigger exporters efficiency or make LabelPairs more efficient in the current NewConst* API. For the latter, we could also try a global (or local) set of `sync.Pool` pools, or something longer term for both LabelPair and LabelPair slices (typically small). It feels collaborating on efficient mutable dto.Metric or something for exporter use has a bigger effect.

> the dto.LabelPair struct contains other struct related to protobuf. Not sure, if they needed. [Prometheus 2.0 does not support protobuf metris scrape](https://github.com/prometheus/client_model/blob/master/README.md), it feels like an expensive leftover.

That's true the LabelPair struct itself could be smaller. However we DO support and intend to have some protobuf metric scrape, so that part has to work (marshalling/unmarshalling). Maybe we could get rid of unknown field support for this type.

**Comment by bwplotka** [comment]
Another solution is along our ideas for generated flow https://github.com/bwplotka/metric-rename-demo/pull/1 - we could create APIs that allow providing LabelPairs directly, so users can keep them ready to use.

**Comment by kgeckhart** [comment]
👋 I ended up in the same area recently looking at further performance improvements to, https://github.com/prometheus-community/yet-another-cloudwatch-exporter/. I put together a PR https://github.com/prometheus/client_golang/pull/1734 that has some positive impacts on resource utilization here. I don't think it's quite enough to say this is "solved" though.
